source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -4752,14 +4832,51 @@ class ActionCollection {
 
                 const allowedActions = this.items.filter(this.isActionAllowed.bind(this));
 
-                for (let i = 0; i < allowedActions.length; i++) {
-                    let actionButton = this.findActionButton(allowedActions[i]);
+                const plainActions: Action[] = [];
+                const overflowActions: Action[] = [];
+                allowedActions.forEach(action => action.mode === Enums.ActionMode.Secondary ? overflowActions.push(action) : plainActions.push(action));
+
+                // given plainActions.length > maxActions, exceeding actions are forced moved to overflow
+                const overflowPrimaryActions = plainActions.splice(hostConfig.actions.maxActions);
+                overflowActions.push(...overflowPrimaryActions);
+
+                const hasOverflow = overflowActions.length > 0;
+
+                let shouldRenderOverflowActionButton = true;
+                if (hasOverflow) {
+                    if (!this.overflowAction) {
+                        this.overflowAction = new OverflowAction(overflowActions);
+                        this.overflowAction.setParent(this._owner);
+                    }
+                    let isRootAction = false;
+                    const card = this.overflowAction.parent?.getRootElement() as AdaptiveCard;
+                    if (card && card.getActionCount() === this.items.length) {
+                        isRootAction = true;
+                        for (let i = 0; i < card.getActionCount(); ++i) {
+                            isRootAction = isRootAction && (card.getActionAt(i) === this.items[i]);
+                        }
+                    }
+                    shouldRenderOverflowActionButton = raiseRenderOverflowActionsEvent(this.overflowAction, isRootAction);
+                }
+
+                const numActionsToRender = plainActions.length + (hasOverflow && shouldRenderOverflowActionButton ? 1 : 0);
```

Comment:
I think this and the below loop could be simplified by simply adding the overflow action to the `mainActions` array as appropriate:
```typescript
if (hasOverflow && shouldRenderOverflowActionButton) {
  mainActions.push(overflowAction);
}

for (let action of mainActions) {
  let actionButton = this.findActionButton(action);

  if (!actionButton) {
    actionButton = new ActionButton(action, parentContainerStyle);
    actionButton.onClick = (ab) => { ab.action.execute(); };

    this.buttons.push(actionButton);
  }

  actionButton.render();

  ...
}
```

Reply:
Cleaned up the logic. Please see the update.